
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>sgkit.call_allele_frequencies</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet">
  <link href="../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../_static/typehints.css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/docsearch.sbt.css" />
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
    <script defer="defer" src="../_static/docsearch.sbt.js"></script>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="sgkit.count_call_alleles" href="sgkit.count_call_alleles.html" />
    <link rel="prev" title="sgkit.save_dataset" href="sgkit.save_dataset.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/sgkit_trnsprnt.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Contents
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../getting_started.html">
   Getting Started
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../user_guide.html">
   User Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../vcf.html">
   Reading VCF
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../examples/index.html">
   Examples
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/gwas_tutorial.html">
     GWAS Tutorial
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../api.html">
   API reference
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.io.bgen.bgen_to_zarr.html">
     sgkit.io.bgen.bgen_to_zarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.io.bgen.read_bgen.html">
     sgkit.io.bgen.read_bgen
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.io.bgen.rechunk_bgen.html">
     sgkit.io.bgen.rechunk_bgen
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.io.plink.read_plink.html">
     sgkit.io.plink.read_plink
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.io.vcf.vcf_to_zarr.html">
     sgkit.io.vcf.vcf_to_zarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.io.vcf.partition_into_regions.html">
     sgkit.io.vcf.partition_into_regions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.io.vcf.vcf_to_zarrs.html">
     sgkit.io.vcf.vcf_to_zarrs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.io.vcf.concat_zarrs.html">
     sgkit.io.vcf.concat_zarrs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.read_scikit_allel_vcfzarr.html">
     sgkit.read_scikit_allel_vcfzarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.load_dataset.html">
     sgkit.load_dataset
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.save_dataset.html">
     sgkit.save_dataset
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     sgkit.call_allele_frequencies
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.count_call_alleles.html">
     sgkit.count_call_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.count_cohort_alleles.html">
     sgkit.count_cohort_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.count_variant_alleles.html">
     sgkit.count_variant_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.divergence.html">
     sgkit.divergence
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.diversity.html">
     sgkit.diversity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.Fst.html">
     sgkit.Fst
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.Garud_H.html">
     sgkit.Garud_H
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.gwas_linear_regression.html">
     sgkit.gwas_linear_regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.hardy_weinberg_test.html">
     sgkit.hardy_weinberg_test
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.identity_by_state.html">
     sgkit.identity_by_state
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.individual_heterozygosity.html">
     sgkit.individual_heterozygosity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.ld_matrix.html">
     sgkit.ld_matrix
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.ld_prune.html">
     sgkit.ld_prune
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.maximal_independent_set.html">
     sgkit.maximal_independent_set
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.observed_heterozygosity.html">
     sgkit.observed_heterozygosity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.pbs.html">
     sgkit.pbs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.pc_relate.html">
     sgkit.pc_relate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.regenie.html">
     sgkit.regenie
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.sample_stats.html">
     sgkit.sample_stats
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.Tajimas_D.html">
     sgkit.Tajimas_D
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variant_stats.html">
     sgkit.variant_stats
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.Weir_Goudet_beta.html">
     sgkit.Weir_Goudet_beta
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.convert_probability_to_call.html">
     sgkit.convert_probability_to_call
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.display_genotypes.html">
     sgkit.display_genotypes
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.filter_partial_calls.html">
     sgkit.filter_partial_calls
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.infer_call_ploidy.html">
     sgkit.infer_call_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.infer_sample_ploidy.html">
     sgkit.infer_sample_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.infer_variant_ploidy.html">
     sgkit.infer_variant_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.simulate_genotype_call_dataset.html">
     sgkit.simulate_genotype_call_dataset
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.window_by_position.html">
     sgkit.window_by_position
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.window_by_variant.html">
     sgkit.window_by_variant
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.regenie_base_prediction_spec.html">
     sgkit.variables.regenie_base_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_allele_count_spec.html">
     sgkit.variables.call_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_allele_frequency_spec.html">
     sgkit.variables.call_allele_frequency_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_dosage_spec.html">
     sgkit.variables.call_dosage_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_dosage_mask_spec.html">
     sgkit.variables.call_dosage_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_genotype_complete_spec.html">
     sgkit.variables.call_genotype_complete_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_genotype_complete_mask_spec.html">
     sgkit.variables.call_genotype_complete_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_genotype_spec.html">
     sgkit.variables.call_genotype_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_genotype_mask_spec.html">
     sgkit.variables.call_genotype_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_genotype_fill_spec.html">
     sgkit.variables.call_genotype_fill_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_genotype_phased_spec.html">
     sgkit.variables.call_genotype_phased_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_genotype_probability_spec.html">
     sgkit.variables.call_genotype_probability_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_genotype_probability_mask_spec.html">
     sgkit.variables.call_genotype_probability_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_heterozygosity_spec.html">
     sgkit.variables.call_heterozygosity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.call_ploidy_spec.html">
     sgkit.variables.call_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.cohort_allele_count_spec.html">
     sgkit.variables.cohort_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.covariates_spec.html">
     sgkit.variables.covariates_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.dosage_spec.html">
     sgkit.variables.dosage_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.genotype_count_spec.html">
     sgkit.variables.genotype_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.ld_prune_index_to_drop_spec.html">
     sgkit.variables.ld_prune_index_to_drop_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.regenie_loco_prediction_spec.html">
     sgkit.variables.regenie_loco_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.regenie_meta_prediction_spec.html">
     sgkit.variables.regenie_meta_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.pc_relate_phi_spec.html">
     sgkit.variables.pc_relate_phi_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_call_rate_spec.html">
     sgkit.variables.sample_call_rate_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_cohort_spec.html">
     sgkit.variables.sample_cohort_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_id_spec.html">
     sgkit.variables.sample_id_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_n_called_spec.html">
     sgkit.variables.sample_n_called_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_n_het_spec.html">
     sgkit.variables.sample_n_het_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_n_hom_alt_spec.html">
     sgkit.variables.sample_n_hom_alt_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_n_hom_ref_spec.html">
     sgkit.variables.sample_n_hom_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_n_non_ref_spec.html">
     sgkit.variables.sample_n_non_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_pca_component_spec.html">
     sgkit.variables.sample_pca_component_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_pca_explained_variance_spec.html">
     sgkit.variables.sample_pca_explained_variance_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_pca_explained_variance_ratio_spec.html">
     sgkit.variables.sample_pca_explained_variance_ratio_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_pca_loading_spec.html">
     sgkit.variables.sample_pca_loading_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_pca_projection_spec.html">
     sgkit.variables.sample_pca_projection_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.sample_ploidy_spec.html">
     sgkit.variables.sample_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_divergence_spec.html">
     sgkit.variables.stat_divergence_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_diversity_spec.html">
     sgkit.variables.stat_diversity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_Fst_spec.html">
     sgkit.variables.stat_Fst_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_Garud_h1_spec.html">
     sgkit.variables.stat_Garud_h1_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_Garud_h12_spec.html">
     sgkit.variables.stat_Garud_h12_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_Garud_h123_spec.html">
     sgkit.variables.stat_Garud_h123_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_Garud_h2_h1_spec.html">
     sgkit.variables.stat_Garud_h2_h1_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_identity_by_state_spec.html">
     sgkit.variables.stat_identity_by_state_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_observed_heterozygosity_spec.html">
     sgkit.variables.stat_observed_heterozygosity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_pbs_spec.html">
     sgkit.variables.stat_pbs_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_Tajimas_D_spec.html">
     sgkit.variables.stat_Tajimas_D_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.stat_Weir_Goudet_beta_spec.html">
     sgkit.variables.stat_Weir_Goudet_beta_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.traits_spec.html">
     sgkit.variables.traits_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_allele_spec.html">
     sgkit.variables.variant_allele_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_allele_count_spec.html">
     sgkit.variables.variant_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_allele_frequency_spec.html">
     sgkit.variables.variant_allele_frequency_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_allele_total_spec.html">
     sgkit.variables.variant_allele_total_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_linreg_beta_spec.html">
     sgkit.variables.variant_linreg_beta_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_call_rate_spec.html">
     sgkit.variables.variant_call_rate_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_contig_spec.html">
     sgkit.variables.variant_contig_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_hwe_p_value_spec.html">
     sgkit.variables.variant_hwe_p_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_id_spec.html">
     sgkit.variables.variant_id_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_n_called_spec.html">
     sgkit.variables.variant_n_called_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_n_het_spec.html">
     sgkit.variables.variant_n_het_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_n_hom_alt_spec.html">
     sgkit.variables.variant_n_hom_alt_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_n_hom_ref_spec.html">
     sgkit.variables.variant_n_hom_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_n_non_ref_spec.html">
     sgkit.variables.variant_n_non_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_linreg_p_value_spec.html">
     sgkit.variables.variant_linreg_p_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_ploidy_spec.html">
     sgkit.variables.variant_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_position_spec.html">
     sgkit.variables.variant_position_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_score_spec.html">
     sgkit.variables.variant_score_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.variant_linreg_t_value_spec.html">
     sgkit.variables.variant_linreg_t_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.window_contig_spec.html">
     sgkit.variables.window_contig_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.window_start_spec.html">
     sgkit.variables.window_start_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sgkit.variables.window_stop_spec.html">
     sgkit.variables.window_stop_spec
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../how_do_i.html">
   How do I …
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../contributing.html">
   Contributing to sgkit
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../about.html">
   About
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/generated/sgkit.call_allele_frequencies.rst.txt"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.rst</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>sgkit.call_allele_frequencies</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="section" id="sgkit-call-allele-frequencies">
<h1>sgkit.call_allele_frequencies<a class="headerlink" href="#sgkit-call-allele-frequencies" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="sgkit.call_allele_frequencies">
<span class="sig-prename descclassname"><span class="pre">sgkit.</span></span><span class="sig-name descname"><span class="pre">call_allele_frequencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ds</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">call_allele_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'call_allele_count'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sgkit.call_allele_frequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute per sample allele frequencies from genotype calls.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ds</strong> : <span class="annotation terse"><a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.20.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dataset</span></code></a></span><span class="annotation full"><a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.20.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dataset</span></code></a></span></dt><dd><p>Dataset containing genotype calls.</p>
</dd>
<dt><strong>call_allele_count</strong> : <span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Hashable" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Hashable</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Hashable" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Hashable</span></code></a></span> (default: <code class="docutils literal notranslate"><span class="pre">'call_allele_count'</span></code>)</dt><dd><p>Input variable name holding call_allele_count as defined by
<a class="reference internal" href="sgkit.variables.call_allele_count_spec.html#sgkit.variables.call_allele_count_spec" title="sgkit.variables.call_allele_count_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.call_allele_count_spec</span></code></a>.
If the variable is not present in <code class="docutils literal notranslate"><span class="pre">ds</span></code>, it will be computed
using <a class="reference internal" href="sgkit.count_call_alleles.html#sgkit.count_call_alleles" title="sgkit.count_call_alleles"><code class="xref py py-func docutils literal notranslate"><span class="pre">count_call_alleles()</span></code></a>.</p>
</dd>
<dt><strong>merge</strong> : <span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span> (default: <code class="docutils literal notranslate"><span class="pre">True</span></code>)</dt><dd><p>If True (the default), merge the input dataset and the computed
output variables into a single dataset, otherwise return only
the computed output variables.
See <a class="reference internal" href="../user_guide.html#dataset-merge"><span class="std std-ref">Dataset merge behavior</span></a> for more details.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><span class="annotation terse"><a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.20.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dataset</span></code></a></span><span class="annotation full"><a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.20.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dataset</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A dataset containing <a class="reference internal" href="sgkit.variables.call_allele_frequency_spec.html#sgkit.variables.call_allele_frequency_spec" title="sgkit.variables.call_allele_frequency_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.call_allele_frequency_spec</span></code></a>
of allele frequencies with shape (variants, samples, alleles) and values
corresponding to the frequency of non-missing occurrences of each allele.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg</span><span class="o">.</span><span class="n">display_genotypes</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span> 
<span class="go">samples    S0   S1</span>
<span class="go">variants</span>
<span class="go">0         1/0  1/0</span>
<span class="go">1         1/0  1/1</span>
<span class="go">2         0/1  1/0</span>
<span class="go">3         0/0  0/0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg</span><span class="o">.</span><span class="n">call_allele_frequencies</span><span class="p">(</span><span class="n">ds</span><span class="p">)[</span><span class="s2">&quot;call_allele_frequency&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> 
<span class="go">array([[[0.5, 0.5],</span>
<span class="go">        [0.5, 0.5]],</span>

<span class="go">       [[0.5, 0.5],</span>
<span class="go">        [0. , 1. ]],</span>

<span class="go">       [[0.5, 0.5],</span>
<span class="go">        [0.5, 0.5]],</span>

<span class="go">       [[1. , 0. ],</span>
<span class="go">        [1. , 0. ]]])</span>
</pre></div>
</div>
</dd></dl>

</div>


              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="sgkit.save_dataset.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">sgkit.save_dataset</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="sgkit.count_call_alleles.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">sgkit.count_call_alleles</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By sgkit developers<br/>
    
        &copy; Copyright 2020, sgkit developers.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>