

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>API reference &mdash; sgkit  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="sgkit.io.plink.read_plink" href="generated/sgkit.io.plink.read_plink.html" />
    <link rel="prev" title="Getting Started" href="getting_started.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> sgkit
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#io-imports">IO/imports</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#plink">PLINK</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/sgkit.io.plink.read_plink.html">sgkit.io.plink.read_plink</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#vcf">VCF</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/sgkit.io.vcf.partition_into_regions.html">sgkit.io.vcf.partition_into_regions</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html">sgkit.io.vcf.vcf_to_zarr</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarrs.html">sgkit.io.vcf.vcf_to_zarrs</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/sgkit.io.vcf.zarrs_to_dataset.html">sgkit.io.vcf.zarrs_to_dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/sgkit.read_vcfzarr.html">sgkit.read_vcfzarr</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#methods">Methods</a><ul>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.count_call_alleles.html">sgkit.count_call_alleles</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.count_cohort_alleles.html">sgkit.count_cohort_alleles</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.count_variant_alleles.html">sgkit.count_variant_alleles</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.divergence.html">sgkit.divergence</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.diversity.html">sgkit.diversity</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.Fst.html">sgkit.Fst</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.Garud_h.html">sgkit.Garud_h</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.gwas_linear_regression.html">sgkit.gwas_linear_regression</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.hardy_weinberg_test.html">sgkit.hardy_weinberg_test</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.regenie.html">sgkit.regenie</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variant_stats.html">sgkit.variant_stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.Tajimas_D.html">sgkit.Tajimas_D</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.pc_relate.html">sgkit.pc_relate</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#utilities">Utilities</a><ul>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.display_genotypes.html">sgkit.display_genotypes</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.simulate_genotype_call_dataset.html">sgkit.simulate_genotype_call_dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.window.html">sgkit.window</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#variables">Variables</a><ul>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.base_prediction_spec.html">sgkit.variables.base_prediction_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.call_allele_count_spec.html">sgkit.variables.call_allele_count_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.call_dosage_spec.html">sgkit.variables.call_dosage_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.call_dosage_mask_spec.html">sgkit.variables.call_dosage_mask_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.call_genotype_spec.html">sgkit.variables.call_genotype_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.call_genotype_mask_spec.html">sgkit.variables.call_genotype_mask_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.call_genotype_phased_spec.html">sgkit.variables.call_genotype_phased_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.call_genotype_probability_spec.html">sgkit.variables.call_genotype_probability_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.call_genotype_probability_mask_spec.html">sgkit.variables.call_genotype_probability_mask_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.covariates_spec.html">sgkit.variables.covariates_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.dosage_spec.html">sgkit.variables.dosage_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.genotype_counts_spec.html">sgkit.variables.genotype_counts_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.loco_prediction_spec.html">sgkit.variables.loco_prediction_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.meta_prediction_spec.html">sgkit.variables.meta_prediction_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.pc_relate_phi_spec.html">sgkit.variables.pc_relate_phi_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.sample_id_spec.html">sgkit.variables.sample_id_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.sample_pcs_spec.html">sgkit.variables.sample_pcs_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.stat_Garud_h1_spec.html">sgkit.variables.stat_Garud_h1_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.stat_Garud_h12_spec.html">sgkit.variables.stat_Garud_h12_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.stat_Garud_h123_spec.html">sgkit.variables.stat_Garud_h123_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.stat_Garud_h2_h1_spec.html">sgkit.variables.stat_Garud_h2_h1_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.traits_spec.html">sgkit.variables.traits_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_allele_spec.html">sgkit.variables.variant_allele_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_allele_count_spec.html">sgkit.variables.variant_allele_count_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_allele_frequency_spec.html">sgkit.variables.variant_allele_frequency_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_allele_total_spec.html">sgkit.variables.variant_allele_total_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_beta_spec.html">sgkit.variables.variant_beta_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_call_rate_spec.html">sgkit.variables.variant_call_rate_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_contig_spec.html">sgkit.variables.variant_contig_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_hwe_p_value_spec.html">sgkit.variables.variant_hwe_p_value_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_id_spec.html">sgkit.variables.variant_id_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_n_called_spec.html">sgkit.variables.variant_n_called_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_n_het_spec.html">sgkit.variables.variant_n_het_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_n_hom_alt_spec.html">sgkit.variables.variant_n_hom_alt_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_n_hom_ref_spec.html">sgkit.variables.variant_n_hom_ref_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_n_non_ref_spec.html">sgkit.variables.variant_n_non_ref_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_p_value_spec.html">sgkit.variables.variant_p_value_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_position_spec.html">sgkit.variables.variant_position_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.variant_t_value_spec.html">sgkit.variables.variant_t_value_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.window_start_spec.html">sgkit.variables.window_start_spec</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/sgkit.variables.window_stop_spec.html">sgkit.variables.window_stop_spec</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="vcf.html">Reading VCF</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to sgkit</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sgkit</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>API reference</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="api-reference">
<h1>API reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h1>
<p>This page provides an auto-generated summary of sgkits’s API.</p>
<div class="section" id="io-imports">
<h2>IO/imports<a class="headerlink" href="#io-imports" title="Permalink to this headline">¶</a></h2>
<div class="section" id="plink">
<h3>PLINK<a class="headerlink" href="#plink" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.io.plink.read_plink.html#sgkit.io.plink.read_plink" title="sgkit.io.plink.read_plink"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_plink</span></code></a>(*[, path, bed_path, bim_path, …])</p></td>
<td><p>Read PLINK dataset.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="vcf">
<h3>VCF<a class="headerlink" href="#vcf" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.io.vcf.partition_into_regions.html#sgkit.io.vcf.partition_into_regions" title="sgkit.io.vcf.partition_into_regions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">partition_into_regions</span></code></a>(vcf_path, *[, …])</p></td>
<td><p>Calculate genomic region strings to partition a compressed VCF or BCF file into roughly equal parts.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">vcf_to_zarr</span></code></a>(input, output, *[, regions, …])</p></td>
<td><p>Convert VCF files to a single Zarr on-disk store.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarrs.html#sgkit.io.vcf.vcf_to_zarrs" title="sgkit.io.vcf.vcf_to_zarrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">vcf_to_zarrs</span></code></a>(input, output, regions[, …])</p></td>
<td><p>Convert VCF files to multiple Zarr on-disk stores, one per region.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.io.vcf.zarrs_to_dataset.html#sgkit.io.vcf.zarrs_to_dataset" title="sgkit.io.vcf.zarrs_to_dataset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">zarrs_to_dataset</span></code></a>(urls[, chunk_length, …])</p></td>
<td><p>Combine multiple Zarr stores into a single Xarray dataset.</p></td>
</tr>
</tbody>
</table>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.read_vcfzarr.html#sgkit.read_vcfzarr" title="sgkit.read_vcfzarr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_vcfzarr</span></code></a>(path)</p></td>
<td><p>Read a VCF Zarr file created using scikit-allel.</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="methods">
<span id="api-methods"></span><h2>Methods<a class="headerlink" href="#methods" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.count_call_alleles.html#sgkit.count_call_alleles" title="sgkit.count_call_alleles"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_call_alleles</span></code></a>(ds, *[, call_genotype, merge])</p></td>
<td><p>Compute per sample allele counts from genotype calls.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.count_cohort_alleles.html#sgkit.count_cohort_alleles" title="sgkit.count_cohort_alleles"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_cohort_alleles</span></code></a>(ds, *[, …])</p></td>
<td><p>Compute per cohort allele counts from per-sample allele counts, or genotype calls.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.count_variant_alleles.html#sgkit.count_variant_alleles" title="sgkit.count_variant_alleles"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_variant_alleles</span></code></a>(ds, *[, …])</p></td>
<td><p>Compute allele count from per-sample allele counts, or genotype calls.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.divergence.html#sgkit.divergence" title="sgkit.divergence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">divergence</span></code></a>(ds, *[, cohort_allele_count, merge])</p></td>
<td><p>Compute divergence between pairs of cohorts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.diversity.html#sgkit.diversity" title="sgkit.diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diversity</span></code></a>(ds, *[, cohort_allele_count, merge])</p></td>
<td><p>Compute diversity from cohort allele counts.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.Fst.html#sgkit.Fst" title="sgkit.Fst"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Fst</span></code></a>(ds, *[, estimator, stat_divergence, merge])</p></td>
<td><p>Compute Fst between pairs of cohorts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.Garud_h.html#sgkit.Garud_h" title="sgkit.Garud_h"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Garud_h</span></code></a>(ds, *[, call_genotype, merge])</p></td>
<td><p>Compute the H1, H12, H123 and H2/H1 statistics for detecting signatures of soft sweeps, as defined in Garud et al.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.gwas_linear_regression.html#sgkit.gwas_linear_regression" title="sgkit.gwas_linear_regression"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gwas_linear_regression</span></code></a>(ds, *, dosage, …[, …])</p></td>
<td><p>Run linear regression to identify continuous trait associations with genetic variants.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.hardy_weinberg_test.html#sgkit.hardy_weinberg_test" title="sgkit.hardy_weinberg_test"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hardy_weinberg_test</span></code></a>(ds, *[, …])</p></td>
<td><p>Exact test for HWE as described in Wigginton et al.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.regenie.html#sgkit.regenie" title="sgkit.regenie"><code class="xref py py-obj docutils literal notranslate"><span class="pre">regenie</span></code></a>(ds, *, dosage, covariates, traits[, …])</p></td>
<td><p>Regenie trait transformation.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variant_stats.html#sgkit.variant_stats" title="sgkit.variant_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variant_stats</span></code></a>(ds, *[, call_genotype_mask, …])</p></td>
<td><p>Compute quality control variant statistics from genotype calls.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.Tajimas_D.html#sgkit.Tajimas_D" title="sgkit.Tajimas_D"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tajimas_D</span></code></a>(ds, *[, variant_allele_count, …])</p></td>
<td><p>Compute Tajimas’ D for a genotype call dataset.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.pc_relate.html#sgkit.pc_relate" title="sgkit.pc_relate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pc_relate</span></code></a>(ds, *[, maf, call_genotype, …])</p></td>
<td><p>Compute PC-Relate as described in Conomos, et al.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="utilities">
<h2>Utilities<a class="headerlink" href="#utilities" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.display_genotypes.html#sgkit.display_genotypes" title="sgkit.display_genotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">display_genotypes</span></code></a>(ds[, max_variants, …])</p></td>
<td><p>Display genotype calls.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.simulate_genotype_call_dataset.html#sgkit.simulate_genotype_call_dataset" title="sgkit.simulate_genotype_call_dataset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">simulate_genotype_call_dataset</span></code></a>(n_variant, …)</p></td>
<td><p>Simulate genotype calls and variant/sample data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.window.html#sgkit.window" title="sgkit.window"><code class="xref py py-obj docutils literal notranslate"><span class="pre">window</span></code></a>(ds, size[, step, merge])</p></td>
<td><p>Add fixed-size windowing information to a dataset.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="variables">
<h2>Variables<a class="headerlink" href="#variables" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.base_prediction_spec.html#sgkit.variables.base_prediction_spec" title="sgkit.variables.base_prediction_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.base_prediction_spec</span></code></a></p></td>
<td><p>REGENIE’s base prediction (blocks, alphas, samples, outcomes).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.call_allele_count_spec.html#sgkit.variables.call_allele_count_spec" title="sgkit.variables.call_allele_count_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.call_allele_count_spec</span></code></a></p></td>
<td><p>Allele counts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.call_dosage_spec.html#sgkit.variables.call_dosage_spec" title="sgkit.variables.call_dosage_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.call_dosage_spec</span></code></a></p></td>
<td><p>Dosages, encoded as floats, with NaN indicating a missing value.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.call_dosage_mask_spec.html#sgkit.variables.call_dosage_mask_spec" title="sgkit.variables.call_dosage_mask_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.call_dosage_mask_spec</span></code></a></p></td>
<td><p>TODO</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.call_genotype_spec.html#sgkit.variables.call_genotype_spec" title="sgkit.variables.call_genotype_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.call_genotype_spec</span></code></a></p></td>
<td><p>Call genotype.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.call_genotype_mask_spec.html#sgkit.variables.call_genotype_mask_spec" title="sgkit.variables.call_genotype_mask_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.call_genotype_mask_spec</span></code></a></p></td>
<td><p>TODO</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.call_genotype_phased_spec.html#sgkit.variables.call_genotype_phased_spec" title="sgkit.variables.call_genotype_phased_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.call_genotype_phased_spec</span></code></a></p></td>
<td><p>A flag for each call indicating if it is phased or not.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.call_genotype_probability_spec.html#sgkit.variables.call_genotype_probability_spec" title="sgkit.variables.call_genotype_probability_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.call_genotype_probability_spec</span></code></a></p></td>
<td><p>TODO</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.call_genotype_probability_mask_spec.html#sgkit.variables.call_genotype_probability_mask_spec" title="sgkit.variables.call_genotype_probability_mask_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.call_genotype_probability_mask_spec</span></code></a></p></td>
<td><p>TODO</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.covariates_spec.html#sgkit.variables.covariates_spec" title="sgkit.variables.covariates_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.covariates_spec</span></code></a></p></td>
<td><p>Covariate variable names.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.dosage_spec.html#sgkit.variables.dosage_spec" title="sgkit.variables.dosage_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.dosage_spec</span></code></a></p></td>
<td><p>Dosage variable name.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.genotype_counts_spec.html#sgkit.variables.genotype_counts_spec" title="sgkit.variables.genotype_counts_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.genotype_counts_spec</span></code></a></p></td>
<td><p>Genotype counts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.loco_prediction_spec.html#sgkit.variables.loco_prediction_spec" title="sgkit.variables.loco_prediction_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.loco_prediction_spec</span></code></a></p></td>
<td><p>REGENIE’s loco_prediction (contigs, samples, outcomes).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.meta_prediction_spec.html#sgkit.variables.meta_prediction_spec" title="sgkit.variables.meta_prediction_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.meta_prediction_spec</span></code></a></p></td>
<td><p>REGENIE’s meta_prediction (samples, outcomes).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.pc_relate_phi_spec.html#sgkit.variables.pc_relate_phi_spec" title="sgkit.variables.pc_relate_phi_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.pc_relate_phi_spec</span></code></a></p></td>
<td><p>PC Relate kinship coefficient matrix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.sample_id_spec.html#sgkit.variables.sample_id_spec" title="sgkit.variables.sample_id_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.sample_id_spec</span></code></a></p></td>
<td><p>The unique identifier of the sample.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.sample_pcs_spec.html#sgkit.variables.sample_pcs_spec" title="sgkit.variables.sample_pcs_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.sample_pcs_spec</span></code></a></p></td>
<td><p>Sample PCs (PCxS).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.stat_Garud_h1_spec.html#sgkit.variables.stat_Garud_h1_spec" title="sgkit.variables.stat_Garud_h1_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.stat_Garud_h1_spec</span></code></a></p></td>
<td><p>Garud H1 statistic for cohorts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.stat_Garud_h12_spec.html#sgkit.variables.stat_Garud_h12_spec" title="sgkit.variables.stat_Garud_h12_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.stat_Garud_h12_spec</span></code></a></p></td>
<td><p>Garud H12 statistic for cohorts.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.stat_Garud_h123_spec.html#sgkit.variables.stat_Garud_h123_spec" title="sgkit.variables.stat_Garud_h123_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.stat_Garud_h123_spec</span></code></a></p></td>
<td><p>Garud H123 statistic for cohorts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.stat_Garud_h2_h1_spec.html#sgkit.variables.stat_Garud_h2_h1_spec" title="sgkit.variables.stat_Garud_h2_h1_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.stat_Garud_h2_h1_spec</span></code></a></p></td>
<td><p>Garud H2/H1 statistic for cohorts.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.traits_spec.html#sgkit.variables.traits_spec" title="sgkit.variables.traits_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.traits_spec</span></code></a></p></td>
<td><p>Trait (for example phenotype) variable names.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_allele_spec.html#sgkit.variables.variant_allele_spec" title="sgkit.variables.variant_allele_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_allele_spec</span></code></a></p></td>
<td><p>The possible alleles for the variant.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_allele_count_spec.html#sgkit.variables.variant_allele_count_spec" title="sgkit.variables.variant_allele_count_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_allele_count_spec</span></code></a></p></td>
<td><p>Variant allele counts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_allele_frequency_spec.html#sgkit.variables.variant_allele_frequency_spec" title="sgkit.variables.variant_allele_frequency_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_allele_frequency_spec</span></code></a></p></td>
<td><p>The frequency of the occurrence of each allele.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_allele_total_spec.html#sgkit.variables.variant_allele_total_spec" title="sgkit.variables.variant_allele_total_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_allele_total_spec</span></code></a></p></td>
<td><p>The number of occurrences of all alleles.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_beta_spec.html#sgkit.variables.variant_beta_spec" title="sgkit.variables.variant_beta_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_beta_spec</span></code></a></p></td>
<td><p>Beta values associated with each variant and trait.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_call_rate_spec.html#sgkit.variables.variant_call_rate_spec" title="sgkit.variables.variant_call_rate_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_call_rate_spec</span></code></a></p></td>
<td><p>The number of samples with heterozygous calls.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_contig_spec.html#sgkit.variables.variant_contig_spec" title="sgkit.variables.variant_contig_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_contig_spec</span></code></a></p></td>
<td><p>Index corresponding to contig name for each variant.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_hwe_p_value_spec.html#sgkit.variables.variant_hwe_p_value_spec" title="sgkit.variables.variant_hwe_p_value_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_hwe_p_value_spec</span></code></a></p></td>
<td><p>P values from HWE test for each variant as float in [0, 1].</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_id_spec.html#sgkit.variables.variant_id_spec" title="sgkit.variables.variant_id_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_id_spec</span></code></a></p></td>
<td><p>The unique identifier of the variant.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_n_called_spec.html#sgkit.variables.variant_n_called_spec" title="sgkit.variables.variant_n_called_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_n_called_spec</span></code></a></p></td>
<td><p>The number of samples with called genotypes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_n_het_spec.html#sgkit.variables.variant_n_het_spec" title="sgkit.variables.variant_n_het_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_n_het_spec</span></code></a></p></td>
<td><p>The number of samples with heterozygous calls.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_n_hom_alt_spec.html#sgkit.variables.variant_n_hom_alt_spec" title="sgkit.variables.variant_n_hom_alt_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_n_hom_alt_spec</span></code></a></p></td>
<td><p>The number of samples with homozygous alternate calls.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_n_hom_ref_spec.html#sgkit.variables.variant_n_hom_ref_spec" title="sgkit.variables.variant_n_hom_ref_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_n_hom_ref_spec</span></code></a></p></td>
<td><p>The number of samples with homozygous reference calls.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_n_non_ref_spec.html#sgkit.variables.variant_n_non_ref_spec" title="sgkit.variables.variant_n_non_ref_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_n_non_ref_spec</span></code></a></p></td>
<td><p>The number of samples that are not homozygous reference calls.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_p_value_spec.html#sgkit.variables.variant_p_value_spec" title="sgkit.variables.variant_p_value_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_p_value_spec</span></code></a></p></td>
<td><p>P values as float in [0, 1].</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_position_spec.html#sgkit.variables.variant_position_spec" title="sgkit.variables.variant_position_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_position_spec</span></code></a></p></td>
<td><p>The reference position of the variant.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.variant_t_value_spec.html#sgkit.variables.variant_t_value_spec" title="sgkit.variables.variant_t_value_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.variant_t_value_spec</span></code></a></p></td>
<td><p>T statistics for each beta.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/sgkit.variables.window_start_spec.html#sgkit.variables.window_start_spec" title="sgkit.variables.window_start_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.window_start_spec</span></code></a></p></td>
<td><p>The index values of window start positions along the <code class="docutils literal notranslate"><span class="pre">variants</span></code> dimension.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/sgkit.variables.window_stop_spec.html#sgkit.variables.window_stop_spec" title="sgkit.variables.window_stop_spec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variables.window_stop_spec</span></code></a></p></td>
<td><p>The index values of window stop positions along the <code class="docutils literal notranslate"><span class="pre">variants</span></code> dimension.</p></td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="generated/sgkit.io.plink.read_plink.html" class="btn btn-neutral float-right" title="sgkit.io.plink.read_plink" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="getting_started.html" class="btn btn-neutral float-left" title="Getting Started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, sgkit developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>