
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Index</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/typehints.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/docsearch.sbt.css" />
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
    <script defer="defer" src="_static/docsearch.sbt.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="#" />
    <link rel="search" title="Search" href="search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/sgkit_trnsprnt.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Contents
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="getting_started.html">
   Getting Started
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="user_guide.html">
   User Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vcf.html">
   Reading VCF
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="examples/index.html">
   Examples
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/gwas_tutorial.html">
     GWAS Tutorial
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="api.html">
   API reference
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.bgen.bgen_to_zarr.html">
     sgkit.io.bgen.bgen_to_zarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.bgen.read_bgen.html">
     sgkit.io.bgen.read_bgen
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.bgen.rechunk_bgen.html">
     sgkit.io.bgen.rechunk_bgen
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.plink.read_plink.html">
     sgkit.io.plink.read_plink
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html">
     sgkit.io.vcf.vcf_to_zarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.vcf.partition_into_regions.html">
     sgkit.io.vcf.partition_into_regions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarrs.html">
     sgkit.io.vcf.vcf_to_zarrs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.vcf.concat_zarrs.html">
     sgkit.io.vcf.concat_zarrs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.read_scikit_allel_vcfzarr.html">
     sgkit.read_scikit_allel_vcfzarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.load_dataset.html">
     sgkit.load_dataset
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.save_dataset.html">
     sgkit.save_dataset
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.additive_relationships.html">
     sgkit.additive_relationships
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.call_allele_frequencies.html">
     sgkit.call_allele_frequencies
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.cohort_allele_frequencies.html">
     sgkit.cohort_allele_frequencies
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.count_call_alleles.html">
     sgkit.count_call_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.count_cohort_alleles.html">
     sgkit.count_cohort_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.count_variant_alleles.html">
     sgkit.count_variant_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.divergence.html">
     sgkit.divergence
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.diversity.html">
     sgkit.diversity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.Fst.html">
     sgkit.Fst
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.Garud_H.html">
     sgkit.Garud_H
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.gwas_linear_regression.html">
     sgkit.gwas_linear_regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.hardy_weinberg_test.html">
     sgkit.hardy_weinberg_test
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.identity_by_state.html">
     sgkit.identity_by_state
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.individual_heterozygosity.html">
     sgkit.individual_heterozygosity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.inverse_additive_relationships.html">
     sgkit.inverse_additive_relationships
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.ld_matrix.html">
     sgkit.ld_matrix
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.ld_prune.html">
     sgkit.ld_prune
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.maximal_independent_set.html">
     sgkit.maximal_independent_set
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.observed_heterozygosity.html">
     sgkit.observed_heterozygosity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.pbs.html">
     sgkit.pbs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.pedigree_kinship.html">
     sgkit.pedigree_kinship
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.pc_relate.html">
     sgkit.pc_relate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.regenie.html">
     sgkit.regenie
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.sample_stats.html">
     sgkit.sample_stats
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.Tajimas_D.html">
     sgkit.Tajimas_D
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variant_stats.html">
     sgkit.variant_stats
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.Weir_Goudet_beta.html">
     sgkit.Weir_Goudet_beta
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.convert_probability_to_call.html">
     sgkit.convert_probability_to_call
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.display_genotypes.html">
     sgkit.display_genotypes
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.filter_partial_calls.html">
     sgkit.filter_partial_calls
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.infer_call_ploidy.html">
     sgkit.infer_call_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.infer_sample_ploidy.html">
     sgkit.infer_sample_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.infer_variant_ploidy.html">
     sgkit.infer_variant_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.parent_indices.html">
     sgkit.parent_indices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.simulate_genotype_call_dataset.html">
     sgkit.simulate_genotype_call_dataset
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.window_by_position.html">
     sgkit.window_by_position
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.window_by_variant.html">
     sgkit.window_by_variant
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.regenie_base_prediction_spec.html">
     sgkit.variables.regenie_base_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_allele_count_spec.html">
     sgkit.variables.call_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_allele_frequency_spec.html">
     sgkit.variables.call_allele_frequency_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_dosage_spec.html">
     sgkit.variables.call_dosage_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_dosage_mask_spec.html">
     sgkit.variables.call_dosage_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_complete_spec.html">
     sgkit.variables.call_genotype_complete_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_complete_mask_spec.html">
     sgkit.variables.call_genotype_complete_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_spec.html">
     sgkit.variables.call_genotype_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_mask_spec.html">
     sgkit.variables.call_genotype_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_fill_spec.html">
     sgkit.variables.call_genotype_fill_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_phased_spec.html">
     sgkit.variables.call_genotype_phased_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_probability_spec.html">
     sgkit.variables.call_genotype_probability_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_probability_mask_spec.html">
     sgkit.variables.call_genotype_probability_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_heterozygosity_spec.html">
     sgkit.variables.call_heterozygosity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_ploidy_spec.html">
     sgkit.variables.call_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.cohort_allele_count_spec.html">
     sgkit.variables.cohort_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.cohort_allele_frequency_spec.html">
     sgkit.variables.cohort_allele_frequency_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.covariates_spec.html">
     sgkit.variables.covariates_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.dosage_spec.html">
     sgkit.variables.dosage_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.genotype_count_spec.html">
     sgkit.variables.genotype_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.ld_prune_index_to_drop_spec.html">
     sgkit.variables.ld_prune_index_to_drop_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.regenie_loco_prediction_spec.html">
     sgkit.variables.regenie_loco_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.regenie_meta_prediction_spec.html">
     sgkit.variables.regenie_meta_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.parent_spec.html">
     sgkit.variables.parent_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.parent_id_spec.html">
     sgkit.variables.parent_id_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.pc_relate_phi_spec.html">
     sgkit.variables.pc_relate_phi_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_call_rate_spec.html">
     sgkit.variables.sample_call_rate_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_cohort_spec.html">
     sgkit.variables.sample_cohort_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_id_spec.html">
     sgkit.variables.sample_id_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_called_spec.html">
     sgkit.variables.sample_n_called_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_het_spec.html">
     sgkit.variables.sample_n_het_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_hom_alt_spec.html">
     sgkit.variables.sample_n_hom_alt_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_hom_ref_spec.html">
     sgkit.variables.sample_n_hom_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_non_ref_spec.html">
     sgkit.variables.sample_n_non_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_component_spec.html">
     sgkit.variables.sample_pca_component_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_explained_variance_spec.html">
     sgkit.variables.sample_pca_explained_variance_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_explained_variance_ratio_spec.html">
     sgkit.variables.sample_pca_explained_variance_ratio_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_loading_spec.html">
     sgkit.variables.sample_pca_loading_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_projection_spec.html">
     sgkit.variables.sample_pca_projection_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_ploidy_spec.html">
     sgkit.variables.sample_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_additive_relationship_spec.html">
     sgkit.variables.stat_additive_relationship_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_divergence_spec.html">
     sgkit.variables.stat_divergence_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_diversity_spec.html">
     sgkit.variables.stat_diversity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Fst_spec.html">
     sgkit.variables.stat_Fst_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Garud_h1_spec.html">
     sgkit.variables.stat_Garud_h1_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Garud_h12_spec.html">
     sgkit.variables.stat_Garud_h12_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Garud_h123_spec.html">
     sgkit.variables.stat_Garud_h123_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Garud_h2_h1_spec.html">
     sgkit.variables.stat_Garud_h2_h1_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Hamilton_Kerr_lambda_spec.html">
     sgkit.variables.stat_Hamilton_Kerr_lambda_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Hamilton_Kerr_tau_spec.html">
     sgkit.variables.stat_Hamilton_Kerr_tau_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_identity_by_state_spec.html">
     sgkit.variables.stat_identity_by_state_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_inverse_additive_relationship_spec.html">
     sgkit.variables.stat_inverse_additive_relationship_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_observed_heterozygosity_spec.html">
     sgkit.variables.stat_observed_heterozygosity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_pbs_spec.html">
     sgkit.variables.stat_pbs_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_pedigree_kinship_spec.html">
     sgkit.variables.stat_pedigree_kinship_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Tajimas_D_spec.html">
     sgkit.variables.stat_Tajimas_D_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Weir_Goudet_beta_spec.html">
     sgkit.variables.stat_Weir_Goudet_beta_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.traits_spec.html">
     sgkit.variables.traits_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_allele_spec.html">
     sgkit.variables.variant_allele_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_allele_count_spec.html">
     sgkit.variables.variant_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_allele_frequency_spec.html">
     sgkit.variables.variant_allele_frequency_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_allele_total_spec.html">
     sgkit.variables.variant_allele_total_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_linreg_beta_spec.html">
     sgkit.variables.variant_linreg_beta_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_call_rate_spec.html">
     sgkit.variables.variant_call_rate_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_contig_spec.html">
     sgkit.variables.variant_contig_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_hwe_p_value_spec.html">
     sgkit.variables.variant_hwe_p_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_id_spec.html">
     sgkit.variables.variant_id_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_called_spec.html">
     sgkit.variables.variant_n_called_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_het_spec.html">
     sgkit.variables.variant_n_het_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_hom_alt_spec.html">
     sgkit.variables.variant_n_hom_alt_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_hom_ref_spec.html">
     sgkit.variables.variant_n_hom_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_non_ref_spec.html">
     sgkit.variables.variant_n_non_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_linreg_p_value_spec.html">
     sgkit.variables.variant_linreg_p_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_ploidy_spec.html">
     sgkit.variables.variant_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_position_spec.html">
     sgkit.variables.variant_position_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_score_spec.html">
     sgkit.variables.variant_score_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_linreg_t_value_spec.html">
     sgkit.variables.variant_linreg_t_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.window_contig_spec.html">
     sgkit.variables.window_contig_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.window_start_spec.html">
     sgkit.variables.window_start_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.window_stop_spec.html">
     sgkit.variables.window_stop_spec
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="how_do_i.html">
   How do I …
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="contributing.html">
   Contributing to sgkit
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="about.html">
   About
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/pystatgen/sgkit"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/pystatgen/sgkit/issues/new?title=Issue%20on%20page%20%2Fgenindex.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1></h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#A"><strong>A</strong></a>
 | <a href="#B"><strong>B</strong></a>
 | <a href="#C"><strong>C</strong></a>
 | <a href="#D"><strong>D</strong></a>
 | <a href="#F"><strong>F</strong></a>
 | <a href="#G"><strong>G</strong></a>
 | <a href="#H"><strong>H</strong></a>
 | <a href="#I"><strong>I</strong></a>
 | <a href="#L"><strong>L</strong></a>
 | <a href="#M"><strong>M</strong></a>
 | <a href="#O"><strong>O</strong></a>
 | <a href="#P"><strong>P</strong></a>
 | <a href="#R"><strong>R</strong></a>
 | <a href="#S"><strong>S</strong></a>
 | <a href="#T"><strong>T</strong></a>
 | <a href="#V"><strong>V</strong></a>
 | <a href="#W"><strong>W</strong></a>
 
</div>
<h2 id="A">A</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.additive_relationships.html#sgkit.additive_relationships">additive_relationships() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="B">B</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.io.bgen.bgen_to_zarr.html#sgkit.io.bgen.bgen_to_zarr">bgen_to_zarr() (in module sgkit.io.bgen)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="C">C</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.call_allele_count_spec.html#sgkit.variables.call_allele_count_spec">call_allele_count_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.call_allele_frequencies.html#sgkit.call_allele_frequencies">call_allele_frequencies() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.variables.call_allele_frequency_spec.html#sgkit.variables.call_allele_frequency_spec">call_allele_frequency_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_dosage_mask_spec.html#sgkit.variables.call_dosage_mask_spec">call_dosage_mask_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_dosage_spec.html#sgkit.variables.call_dosage_spec">call_dosage_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_genotype_complete_mask_spec.html#sgkit.variables.call_genotype_complete_mask_spec">call_genotype_complete_mask_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_genotype_complete_spec.html#sgkit.variables.call_genotype_complete_spec">call_genotype_complete_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_genotype_fill_spec.html#sgkit.variables.call_genotype_fill_spec">call_genotype_fill_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_genotype_mask_spec.html#sgkit.variables.call_genotype_mask_spec">call_genotype_mask_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_genotype_phased_spec.html#sgkit.variables.call_genotype_phased_spec">call_genotype_phased_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_genotype_probability_mask_spec.html#sgkit.variables.call_genotype_probability_mask_spec">call_genotype_probability_mask_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_genotype_probability_spec.html#sgkit.variables.call_genotype_probability_spec">call_genotype_probability_spec (in module sgkit.variables)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.call_genotype_spec.html#sgkit.variables.call_genotype_spec">call_genotype_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_heterozygosity_spec.html#sgkit.variables.call_heterozygosity_spec">call_heterozygosity_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.call_ploidy_spec.html#sgkit.variables.call_ploidy_spec">call_ploidy_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.cohort_allele_count_spec.html#sgkit.variables.cohort_allele_count_spec">cohort_allele_count_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.cohort_allele_frequencies.html#sgkit.cohort_allele_frequencies">cohort_allele_frequencies() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.variables.cohort_allele_frequency_spec.html#sgkit.variables.cohort_allele_frequency_spec">cohort_allele_frequency_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.io.vcf.concat_zarrs.html#sgkit.io.vcf.concat_zarrs">concat_zarrs() (in module sgkit.io.vcf)</a>
</li>
      <li><a href="generated/sgkit.convert_probability_to_call.html#sgkit.convert_probability_to_call">convert_probability_to_call() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.count_call_alleles.html#sgkit.count_call_alleles">count_call_alleles() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.count_cohort_alleles.html#sgkit.count_cohort_alleles">count_cohort_alleles() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.count_variant_alleles.html#sgkit.count_variant_alleles">count_variant_alleles() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.variables.covariates_spec.html#sgkit.variables.covariates_spec">covariates_spec (in module sgkit.variables)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="D">D</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.display_genotypes.html#sgkit.display_genotypes">display_genotypes() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.divergence.html#sgkit.divergence">divergence() (in module sgkit)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.diversity.html#sgkit.diversity">diversity() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.variables.dosage_spec.html#sgkit.variables.dosage_spec">dosage_spec (in module sgkit.variables)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="F">F</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.filter_partial_calls.html#sgkit.filter_partial_calls">filter_partial_calls() (in module sgkit)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.Fst.html#sgkit.Fst">Fst() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="G">G</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.Garud_H.html#sgkit.Garud_H">Garud_H() (in module sgkit)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.genotype_count_spec.html#sgkit.variables.genotype_count_spec">genotype_count_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.gwas_linear_regression.html#sgkit.gwas_linear_regression">gwas_linear_regression() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="H">H</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.hardy_weinberg_test.html#sgkit.hardy_weinberg_test">hardy_weinberg_test() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="I">I</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.identity_by_state.html#sgkit.identity_by_state">identity_by_state() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.individual_heterozygosity.html#sgkit.individual_heterozygosity">individual_heterozygosity() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.infer_call_ploidy.html#sgkit.infer_call_ploidy">infer_call_ploidy() (in module sgkit)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.infer_sample_ploidy.html#sgkit.infer_sample_ploidy">infer_sample_ploidy() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.infer_variant_ploidy.html#sgkit.infer_variant_ploidy">infer_variant_ploidy() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.inverse_additive_relationships.html#sgkit.inverse_additive_relationships">inverse_additive_relationships() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="L">L</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.ld_matrix.html#sgkit.ld_matrix">ld_matrix() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.ld_prune.html#sgkit.ld_prune">ld_prune() (in module sgkit)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.ld_prune_index_to_drop_spec.html#sgkit.variables.ld_prune_index_to_drop_spec">ld_prune_index_to_drop_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.load_dataset.html#sgkit.load_dataset">load_dataset() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="M">M</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.maximal_independent_set.html#sgkit.maximal_independent_set">maximal_independent_set() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="O">O</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.observed_heterozygosity.html#sgkit.observed_heterozygosity">observed_heterozygosity() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.parent_id_spec.html#sgkit.variables.parent_id_spec">parent_id_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.parent_indices.html#sgkit.parent_indices">parent_indices() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.variables.parent_spec.html#sgkit.variables.parent_spec">parent_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.io.vcf.partition_into_regions.html#sgkit.io.vcf.partition_into_regions">partition_into_regions() (in module sgkit.io.vcf)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.pbs.html#sgkit.pbs">pbs() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.pc_relate.html#sgkit.pc_relate">pc_relate() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.variables.pc_relate_phi_spec.html#sgkit.variables.pc_relate_phi_spec">pc_relate_phi_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.pedigree_kinship.html#sgkit.pedigree_kinship">pedigree_kinship() (in module sgkit)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="R">R</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.io.bgen.read_bgen.html#sgkit.io.bgen.read_bgen">read_bgen() (in module sgkit.io.bgen)</a>
</li>
      <li><a href="generated/sgkit.io.plink.read_plink.html#sgkit.io.plink.read_plink">read_plink() (in module sgkit.io.plink)</a>
</li>
      <li><a href="generated/sgkit.read_scikit_allel_vcfzarr.html#sgkit.read_scikit_allel_vcfzarr">read_scikit_allel_vcfzarr() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.io.bgen.rechunk_bgen.html#sgkit.io.bgen.rechunk_bgen">rechunk_bgen() (in module sgkit.io.bgen)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.regenie.html#sgkit.regenie">regenie() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.variables.regenie_base_prediction_spec.html#sgkit.variables.regenie_base_prediction_spec">regenie_base_prediction_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.regenie_loco_prediction_spec.html#sgkit.variables.regenie_loco_prediction_spec">regenie_loco_prediction_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.regenie_meta_prediction_spec.html#sgkit.variables.regenie_meta_prediction_spec">regenie_meta_prediction_spec (in module sgkit.variables)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="S">S</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.sample_call_rate_spec.html#sgkit.variables.sample_call_rate_spec">sample_call_rate_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_cohort_spec.html#sgkit.variables.sample_cohort_spec">sample_cohort_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_id_spec.html#sgkit.variables.sample_id_spec">sample_id_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_n_called_spec.html#sgkit.variables.sample_n_called_spec">sample_n_called_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_n_het_spec.html#sgkit.variables.sample_n_het_spec">sample_n_het_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_n_hom_alt_spec.html#sgkit.variables.sample_n_hom_alt_spec">sample_n_hom_alt_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_n_hom_ref_spec.html#sgkit.variables.sample_n_hom_ref_spec">sample_n_hom_ref_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_n_non_ref_spec.html#sgkit.variables.sample_n_non_ref_spec">sample_n_non_ref_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_pca_component_spec.html#sgkit.variables.sample_pca_component_spec">sample_pca_component_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_pca_explained_variance_ratio_spec.html#sgkit.variables.sample_pca_explained_variance_ratio_spec">sample_pca_explained_variance_ratio_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_pca_explained_variance_spec.html#sgkit.variables.sample_pca_explained_variance_spec">sample_pca_explained_variance_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_pca_loading_spec.html#sgkit.variables.sample_pca_loading_spec">sample_pca_loading_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_pca_projection_spec.html#sgkit.variables.sample_pca_projection_spec">sample_pca_projection_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.sample_ploidy_spec.html#sgkit.variables.sample_ploidy_spec">sample_ploidy_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.sample_stats.html#sgkit.sample_stats">sample_stats() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.save_dataset.html#sgkit.save_dataset">save_dataset() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.simulate_genotype_call_dataset.html#sgkit.simulate_genotype_call_dataset">simulate_genotype_call_dataset() (in module sgkit)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.stat_additive_relationship_spec.html#sgkit.variables.stat_additive_relationship_spec">stat_additive_relationship_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_divergence_spec.html#sgkit.variables.stat_divergence_spec">stat_divergence_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_diversity_spec.html#sgkit.variables.stat_diversity_spec">stat_diversity_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Fst_spec.html#sgkit.variables.stat_Fst_spec">stat_Fst_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Garud_h123_spec.html#sgkit.variables.stat_Garud_h123_spec">stat_Garud_h123_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Garud_h12_spec.html#sgkit.variables.stat_Garud_h12_spec">stat_Garud_h12_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Garud_h1_spec.html#sgkit.variables.stat_Garud_h1_spec">stat_Garud_h1_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Garud_h2_h1_spec.html#sgkit.variables.stat_Garud_h2_h1_spec">stat_Garud_h2_h1_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Hamilton_Kerr_lambda_spec.html#sgkit.variables.stat_Hamilton_Kerr_lambda_spec">stat_Hamilton_Kerr_lambda_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Hamilton_Kerr_tau_spec.html#sgkit.variables.stat_Hamilton_Kerr_tau_spec">stat_Hamilton_Kerr_tau_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_identity_by_state_spec.html#sgkit.variables.stat_identity_by_state_spec">stat_identity_by_state_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_inverse_additive_relationship_spec.html#sgkit.variables.stat_inverse_additive_relationship_spec">stat_inverse_additive_relationship_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_observed_heterozygosity_spec.html#sgkit.variables.stat_observed_heterozygosity_spec">stat_observed_heterozygosity_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_pbs_spec.html#sgkit.variables.stat_pbs_spec">stat_pbs_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_pedigree_kinship_spec.html#sgkit.variables.stat_pedigree_kinship_spec">stat_pedigree_kinship_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Tajimas_D_spec.html#sgkit.variables.stat_Tajimas_D_spec">stat_Tajimas_D_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.stat_Weir_Goudet_beta_spec.html#sgkit.variables.stat_Weir_Goudet_beta_spec">stat_Weir_Goudet_beta_spec (in module sgkit.variables)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="T">T</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.Tajimas_D.html#sgkit.Tajimas_D">Tajimas_D() (in module sgkit)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.traits_spec.html#sgkit.variables.traits_spec">traits_spec (in module sgkit.variables)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="V">V</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.variant_allele_count_spec.html#sgkit.variables.variant_allele_count_spec">variant_allele_count_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_allele_frequency_spec.html#sgkit.variables.variant_allele_frequency_spec">variant_allele_frequency_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_allele_spec.html#sgkit.variables.variant_allele_spec">variant_allele_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_allele_total_spec.html#sgkit.variables.variant_allele_total_spec">variant_allele_total_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_call_rate_spec.html#sgkit.variables.variant_call_rate_spec">variant_call_rate_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_contig_spec.html#sgkit.variables.variant_contig_spec">variant_contig_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_hwe_p_value_spec.html#sgkit.variables.variant_hwe_p_value_spec">variant_hwe_p_value_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_id_spec.html#sgkit.variables.variant_id_spec">variant_id_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_linreg_beta_spec.html#sgkit.variables.variant_linreg_beta_spec">variant_linreg_beta_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_linreg_p_value_spec.html#sgkit.variables.variant_linreg_p_value_spec">variant_linreg_p_value_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_linreg_t_value_spec.html#sgkit.variables.variant_linreg_t_value_spec">variant_linreg_t_value_spec (in module sgkit.variables)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.variant_n_called_spec.html#sgkit.variables.variant_n_called_spec">variant_n_called_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_n_het_spec.html#sgkit.variables.variant_n_het_spec">variant_n_het_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_n_hom_alt_spec.html#sgkit.variables.variant_n_hom_alt_spec">variant_n_hom_alt_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_n_hom_ref_spec.html#sgkit.variables.variant_n_hom_ref_spec">variant_n_hom_ref_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_n_non_ref_spec.html#sgkit.variables.variant_n_non_ref_spec">variant_n_non_ref_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_ploidy_spec.html#sgkit.variables.variant_ploidy_spec">variant_ploidy_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_position_spec.html#sgkit.variables.variant_position_spec">variant_position_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.variant_score_spec.html#sgkit.variables.variant_score_spec">variant_score_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variant_stats.html#sgkit.variant_stats">variant_stats() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr">vcf_to_zarr() (in module sgkit.io.vcf)</a>
</li>
      <li><a href="generated/sgkit.io.vcf.vcf_to_zarrs.html#sgkit.io.vcf.vcf_to_zarrs">vcf_to_zarrs() (in module sgkit.io.vcf)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="W">W</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.Weir_Goudet_beta.html#sgkit.Weir_Goudet_beta">Weir_Goudet_beta() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.window_by_position.html#sgkit.window_by_position">window_by_position() (in module sgkit)</a>
</li>
      <li><a href="generated/sgkit.window_by_variant.html#sgkit.window_by_variant">window_by_variant() (in module sgkit)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="generated/sgkit.variables.window_contig_spec.html#sgkit.variables.window_contig_spec">window_contig_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.window_start_spec.html#sgkit.variables.window_start_spec">window_start_spec (in module sgkit.variables)</a>
</li>
      <li><a href="generated/sgkit.variables.window_stop_spec.html#sgkit.variables.window_stop_spec">window_stop_spec (in module sgkit.variables)</a>
</li>
  </ul></td>
</tr></table>



              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By sgkit developers<br/>
  
      &copy; Copyright 2020, sgkit developers.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>