

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>User Guide &mdash; sgkit  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Reading VCF" href="vcf.html" />
    <link rel="prev" title="Getting Started" href="getting_started.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/sgkit_white_trnsprnt.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#reading-genetic-data">Reading genetic data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#converting-genetic-data-to-zarr">Converting genetic data to Zarr</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bgen">BGEN</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plink">PLINK</a></li>
<li class="toctree-l3"><a class="reference internal" href="#vcf">VCF</a></li>
<li class="toctree-l3"><a class="reference internal" href="#working-with-cloud-native-data">Working with cloud-native data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#datasets">Datasets</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#dataset-merge-behavior">Dataset merge behavior</a></li>
<li class="toctree-l3"><a class="reference internal" href="#custom-naming-conventions">Custom naming conventions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-custom-data-to-a-dataset">Adding custom data to a Dataset</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#methods">Methods</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#custom-computations">Custom Computations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#pca">PCA</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#deployment">Deployment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#deploying-sgkit-on-a-cluster">Deploying sgkit on a cluster</a></li>
<li class="toctree-l3"><a class="reference internal" href="#using-gpus">Using GPUs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#troubleshooting">Troubleshooting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#monitoring-operations">Monitoring operations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualizing-computations">Visualizing computations</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="vcf.html">Reading VCF</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to sgkit</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sgkit</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>User Guide</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/user_guide.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="user-guide">
<h1>User Guide<a class="headerlink" href="#user-guide" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="table-of-contents">
<p class="topic-title">Table of contents:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#reading-genetic-data" id="id3">Reading genetic data</a></p>
<ul>
<li><p><a class="reference internal" href="#installation" id="id4">Installation</a></p></li>
<li><p><a class="reference internal" href="#converting-genetic-data-to-zarr" id="id5">Converting genetic data to Zarr</a></p></li>
<li><p><a class="reference internal" href="#bgen" id="id6">BGEN</a></p></li>
<li><p><a class="reference internal" href="#plink" id="id7">PLINK</a></p></li>
<li><p><a class="reference internal" href="#vcf" id="id8">VCF</a></p></li>
<li><p><a class="reference internal" href="#working-with-cloud-native-data" id="id9">Working with cloud-native data</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#datasets" id="id10">Datasets</a></p>
<ul>
<li><p><a class="reference internal" href="#dataset-merge-behavior" id="id11">Dataset merge behavior</a></p></li>
<li><p><a class="reference internal" href="#custom-naming-conventions" id="id12">Custom naming conventions</a></p></li>
<li><p><a class="reference internal" href="#adding-custom-data-to-a-dataset" id="id13">Adding custom data to a Dataset</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#methods" id="id14">Methods</a></p>
<ul>
<li><p><a class="reference internal" href="#custom-computations" id="id15">Custom Computations</a></p></li>
<li><p><a class="reference internal" href="#pca" id="id16">PCA</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#deployment" id="id17">Deployment</a></p>
<ul>
<li><p><a class="reference internal" href="#deploying-sgkit-on-a-cluster" id="id18">Deploying sgkit on a cluster</a></p></li>
<li><p><a class="reference internal" href="#using-gpus" id="id19">Using GPUs</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#troubleshooting" id="id20">Troubleshooting</a></p>
<ul>
<li><p><a class="reference internal" href="#monitoring-operations" id="id21">Monitoring operations</a></p></li>
<li><p><a class="reference internal" href="#visualizing-computations" id="id22">Visualizing computations</a></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="reading-genetic-data">
<span id="id1"></span><h2><a class="toc-backref" href="#id3">Reading genetic data</a><a class="headerlink" href="#reading-genetic-data" title="Permalink to this headline">¶</a></h2>
<div class="section" id="installation">
<h3><a class="toc-backref" href="#id4">Installation</a><a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h3>
<p>Sgkit can read standard genetic file formats, including VCF, PLINK, and BGEN. Support for reading
these formats is not installed by default, and requires additional dependencies, which can be installed
as an “extra” feature using pip, as follows.</p>
<p>To install sgkit with BGEN support:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install &#39;sgkit[bgen]&#39;
</pre></div>
</div>
<p>To install sgkit with PLINK support:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install &#39;sgkit[plink]&#39;
</pre></div>
</div>
<p>To install sgkit with VCF support:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install &#39;sgkit[vcf]&#39;
</pre></div>
</div>
</div>
<div class="section" id="converting-genetic-data-to-zarr">
<h3><a class="toc-backref" href="#id5">Converting genetic data to Zarr</a><a class="headerlink" href="#converting-genetic-data-to-zarr" title="Permalink to this headline">¶</a></h3>
<p>There are broadly two ways of loading genetic file format <code class="docutils literal notranslate"><span class="pre">X</span></code> for use in sgkit:</p>
<ol class="arabic simple">
<li><p>a <code class="docutils literal notranslate"><span class="pre">read_X</span></code> function to read the file directly</p></li>
<li><p>a <code class="docutils literal notranslate"><span class="pre">X_to_zarr</span></code> function to convert to Zarr first</p></li>
</ol>
<p>Which one to use depends on the size of the input, as well as the file format (not all file
formats have both options).</p>
<p>Generally speaking, the <code class="docutils literal notranslate"><span class="pre">read_X</span></code> functions are convenience functions that are suitable
for small input files, or where the cost of conversion is small. The <code class="docutils literal notranslate"><span class="pre">X_to_zarr</span></code> functions
are more appropriate for large datasets, since the expensive conversion step need only be
paid once.</p>
<p>After converting a file to Zarr format, it can be loaded with sgkit’s <a class="reference internal" href="generated/sgkit.load_dataset.html#sgkit.load_dataset" title="sgkit.load_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.load_dataset()</span></code></a>.
This is like Xarray’s <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.open_zarr.html#xarray.open_zarr" title="(in xarray v0.17.1.dev0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">xarray.open_zarr()</span></code></a> function, with some useful defaults applied for you.</p>
<p>Similarly, the function <a class="reference internal" href="generated/sgkit.save_dataset.html#sgkit.save_dataset" title="sgkit.save_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.save_dataset()</span></code></a> can be used to save the dataset in Zarr format.
This can be used to save a newly-loaded dataset, or a dataset with new variables that are the
result of running genetic methods on the data, so it can be loaded again later.</p>
</div>
<div class="section" id="bgen">
<h3><a class="toc-backref" href="#id6">BGEN</a><a class="headerlink" href="#bgen" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="generated/sgkit.io.bgen.read_bgen.html#sgkit.io.bgen.read_bgen" title="sgkit.io.bgen.read_bgen"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.bgen.read_bgen()</span></code></a> function loads a single BGEN dataset as Dask
arrays within an <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.17.1.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> from a <code class="docutils literal notranslate"><span class="pre">bgen</span></code> file.</p>
<p>The <a class="reference internal" href="generated/sgkit.io.bgen.bgen_to_zarr.html#sgkit.io.bgen.bgen_to_zarr" title="sgkit.io.bgen.bgen_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.bgen.bgen_to_zarr()</span></code></a> function converts a <code class="docutils literal notranslate"><span class="pre">bgen</span></code> file to Zarr
files stored in sgkit’s Xarray data representation, which can then be opened
as a <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.17.1.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a>.</p>
</div>
<div class="section" id="plink">
<h3><a class="toc-backref" href="#id7">PLINK</a><a class="headerlink" href="#plink" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="generated/sgkit.io.plink.read_plink.html#sgkit.io.plink.read_plink" title="sgkit.io.plink.read_plink"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.plink.read_plink()</span></code></a> function loads a single PLINK dataset as Dask
arrays within an <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.17.1.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> from <code class="docutils literal notranslate"><span class="pre">bed</span></code>, <code class="docutils literal notranslate"><span class="pre">bim</span></code>, and <code class="docutils literal notranslate"><span class="pre">fam</span></code> files.</p>
</div>
<div class="section" id="vcf">
<h3><a class="toc-backref" href="#id8">VCF</a><a class="headerlink" href="#vcf" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a> function converts one or more VCF files to
Zarr files stored in sgkit’s Xarray data representation, which can then be opened
as a <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.17.1.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a>.</p>
<p>See <a class="reference internal" href="vcf.html#vcf"><span class="std std-ref">Reading VCF</span></a> for installation instructions, and details on using VCF in sgkit.</p>
</div>
<div class="section" id="working-with-cloud-native-data">
<h3><a class="toc-backref" href="#id9">Working with cloud-native data</a><a class="headerlink" href="#working-with-cloud-native-data" title="Permalink to this headline">¶</a></h3>
<p>TODO: Show how to read/write Zarr (and VCF?) data in cloud storage</p>
</div>
</div>
<div class="section" id="datasets">
<h2><a class="toc-backref" href="#id10">Datasets</a><a class="headerlink" href="#datasets" title="Permalink to this headline">¶</a></h2>
<div class="section" id="dataset-merge-behavior">
<span id="dataset-merge"></span><h3><a class="toc-backref" href="#id11">Dataset merge behavior</a><a class="headerlink" href="#dataset-merge-behavior" title="Permalink to this headline">¶</a></h3>
<p>Generally, method functions in sgkit compute some new variables based on the
input dataset, then return a new output dataset that consists of the input
dataset plus the new computed variables. The input dataset is unchanged.</p>
<p>This behavior can be controlled using the <code class="docutils literal notranslate"><span class="pre">merge</span></code> parameter. If set to <code class="docutils literal notranslate"><span class="pre">True</span></code>
(the default), then the function will merge the input dataset and the computed
output variables into a single dataset. Output variables will overwrite any
input variables with the same name, and a warning will be issued in this case.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, the function will return only the computed output variables.</p>
<p>Examples:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>

<span class="gp">In [2]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=.</span><span class="mi">1</span><span class="p">)</span>

<span class="gp">In [3]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">ds</span><span class="p">[[</span><span class="s1">&#39;variant_allele&#39;</span><span class="p">,</span> <span class="s1">&#39;call_genotype&#39;</span><span class="p">]]</span>

<span class="gp">In [4]: </span><span class="n">ds</span>
<span class="gh">Out[4]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:         (alleles: 2, ploidy: 2, samples: 50, variants: 100)</span>
<span class="go">Dimensions without coordinates: alleles, ploidy, samples, variants</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele  (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    call_genotype   (variants, samples, ploidy) int8 0 0 1 0 1 0 ... 0 1 0 1 0 0</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0]</span>

<span class="go"># By default, new variables are merged into a copy of the provided dataset</span>
<span class="gp">In [5]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [6]: </span><span class="n">ds</span>
<span class="gh">Out[6]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:               (alleles: 2, ploidy: 2, samples: 50, variants: 100)</span>
<span class="go">Dimensions without coordinates: alleles, ploidy, samples, variants</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele_count  (variants, alleles) uint64 dask.array&lt;chunksize=(100, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_allele_count     (variants, samples, alleles) uint8 dask.array&lt;chunksize=(100, 50, 2), meta=np.ndarray&gt;</span>
<span class="go">    variant_allele        (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    call_genotype         (variants, samples, ploidy) int8 0 0 1 0 1 ... 0 1 0 0</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0]</span>

<span class="go"># If an existing variable would be re-defined, a warning is thrown</span>
<span class="gp">In [7]: </span><span class="kn">import</span> <span class="nn">warnings</span>

<span class="gp">In [8]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [9]: </span><span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">(</span><span class="n">record</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">as</span> <span class="n">w</span><span class="p">:</span>
<span class="gp">   ...: </span>    <span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gp">   ...: </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">w</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">category</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">w</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">message</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">   ...: </span>
<span class="go">MergeWarning: The following variables in the input dataset will be replaced in the output: variant_allele_count</span>

<span class="go"># New variables can also be returned in their own dataset</span>
<span class="gp">In [10]: </span><span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gh">Out[10]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:               (alleles: 2, variants: 100)</span>
<span class="go">Dimensions without coordinates: alleles, variants</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele_count  (variants, alleles) uint64 dask.array&lt;chunksize=(100, 2), meta=np.ndarray&gt;</span>

<span class="go"># This can be useful for merging multiple datasets manually</span>
<span class="gp">In [11]: </span><span class="n">ds</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="gh">Out[11]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:               (alleles: 2, ploidy: 2, samples: 50, variants: 100)</span>
<span class="go">Dimensions without coordinates: alleles, ploidy, samples, variants</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele_count  (variants, alleles) uint64 dask.array&lt;chunksize=(100, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_allele_count     (variants, samples, alleles) uint8 dask.array&lt;chunksize=(100, 50, 2), meta=np.ndarray&gt;</span>
<span class="go">    variant_allele        (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    call_genotype         (variants, samples, ploidy) int8 0 0 1 0 1 ... 0 1 0 0</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0]</span>
</pre></div>
</div>
</div>
<div class="section" id="custom-naming-conventions">
<h3><a class="toc-backref" href="#id12">Custom naming conventions</a><a class="headerlink" href="#custom-naming-conventions" title="Permalink to this headline">¶</a></h3>
<p>TODO: Show to use a custom naming convention via Xarray renaming features.</p>
</div>
<div class="section" id="adding-custom-data-to-a-dataset">
<h3><a class="toc-backref" href="#id13">Adding custom data to a Dataset</a><a class="headerlink" href="#adding-custom-data-to-a-dataset" title="Permalink to this headline">¶</a></h3>
<p>TODO:  Show how something like sample metadata can be joined to an existing Xarray dataset. Also briefly explain
indexing and uniqueness within Xarray/Pandas, since this is critical for understanding joins.</p>
</div>
</div>
<div class="section" id="methods">
<h2><a class="toc-backref" href="#id14">Methods</a><a class="headerlink" href="#methods" title="Permalink to this headline">¶</a></h2>
<div class="section" id="custom-computations">
<span id="id2"></span><h3><a class="toc-backref" href="#id15">Custom Computations</a><a class="headerlink" href="#custom-computations" title="Permalink to this headline">¶</a></h3>
<p>TODO: Finish explaining how Numba works and how users might apply it</p>
<p>Here is an example that demonstrates an alt allele count:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [12]: </span><span class="kn">import</span> <span class="nn">numba</span>

<span class="gp">In [13]: </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>

<span class="gp">In [14]: </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="gp">In [15]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=.</span><span class="mi">2</span><span class="p">)</span>

<span class="gp">In [16]: </span><span class="k">def</span> <span class="nf">alt_allele_count</span><span class="p">(</span><span class="n">gt</span><span class="p">):</span>
<span class="gp">   ....: </span>    <span class="n">out</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">shape</span><span class="p">[:</span><span class="mi">2</span><span class="p">],</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
<span class="gp">   ....: </span>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">ndindex</span><span class="p">(</span><span class="o">*</span><span class="n">out</span><span class="o">.</span><span class="n">shape</span><span class="p">):</span>
<span class="gp">   ....: </span>        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">gt</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">):</span>
<span class="gp">   ....: </span>            <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">gt</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
<span class="gp">   ....: </span>    <span class="k">return</span> <span class="n">out</span>
<span class="gp">   ....: </span>

<span class="gp">In [17]: </span><span class="n">numba</span><span class="o">.</span><span class="n">njit</span><span class="p">(</span><span class="n">alt_allele_count</span><span class="p">)(</span><span class="n">ds</span><span class="o">.</span><span class="n">call_genotype</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="gh">Out[17]: </span><span class="go"></span>
<span class="go">array([[ 0,  1,  1],</span>
<span class="go">       [ 1,  1, -1],</span>
<span class="go">       [-1,  1,  2],</span>
<span class="go">       [ 0, -1, -1],</span>
<span class="go">       [ 1,  2,  2]])</span>
</pre></div>
</div>
</div>
<div class="section" id="pca">
<h3><a class="toc-backref" href="#id16">PCA</a><a class="headerlink" href="#pca" title="Permalink to this headline">¶</a></h3>
<p>TODO: Describe the upstream tools for PCA (i.e. those in dask-ml/scikit-learn)</p>
</div>
</div>
<div class="section" id="deployment">
<h2><a class="toc-backref" href="#id17">Deployment</a><a class="headerlink" href="#deployment" title="Permalink to this headline">¶</a></h2>
<div class="section" id="deploying-sgkit-on-a-cluster">
<h3><a class="toc-backref" href="#id18">Deploying sgkit on a cluster</a><a class="headerlink" href="#deploying-sgkit-on-a-cluster" title="Permalink to this headline">¶</a></h3>
<p>TODO: Create a tutorial on running sgkit at scale</p>
</div>
<div class="section" id="using-gpus">
<h3><a class="toc-backref" href="#id19">Using GPUs</a><a class="headerlink" href="#using-gpus" title="Permalink to this headline">¶</a></h3>
<p>TODO: Show CuPy examples</p>
</div>
</div>
<div class="section" id="troubleshooting">
<h2><a class="toc-backref" href="#id20">Troubleshooting</a><a class="headerlink" href="#troubleshooting" title="Permalink to this headline">¶</a></h2>
<div class="section" id="monitoring-operations">
<h3><a class="toc-backref" href="#id21">Monitoring operations</a><a class="headerlink" href="#monitoring-operations" title="Permalink to this headline">¶</a></h3>
<p>The simplest way to monitor operations when running sgkit on a single host is to use <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-local.html">Dask local diagnostics</a>.</p>
<p>As an example, this code shows how to track the progress of a single sgkit function:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [18]: </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>

<span class="gp">In [19]: </span><span class="kn">from</span> <span class="nn">dask.diagnostics</span> <span class="kn">import</span> <span class="n">ProgressBar</span>

<span class="gp">In [20]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=.</span><span class="mi">1</span><span class="p">)</span>

<span class="gp">In [21]: </span><span class="k">with</span> <span class="n">ProgressBar</span><span class="p">():</span>
<span class="gp">   ....: </span>    <span class="n">ac</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span><span class="o">.</span><span class="n">variant_allele_count</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="gp">   ....: </span>

<span class="go">[                                        ] | 0% Completed |  0.0s</span>
<span class="go">[########################################] | 100% Completed |  0.1s</span>

<span class="gp">In [22]: </span><span class="n">ac</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
<span class="gh">Out[22]: </span><span class="go"></span>
<span class="go">&lt;xarray.DataArray &#39;variant_allele_count&#39; (variants: 5, alleles: 2)&gt;</span>
<span class="go">array([[50, 40],</span>
<span class="go">       [40, 52],</span>
<span class="go">       [48, 45],</span>
<span class="go">       [37, 51],</span>
<span class="go">       [45, 46]], dtype=uint64)</span>
<span class="go">Dimensions without coordinates: variants, alleles</span>
<span class="go">Attributes:</span>
<span class="go">    comment:  Variant allele counts. With shape (variants, alleles) and value...</span>
</pre></div>
</div>
<p>Monitoring resource utilization with <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-local.html#resourceprofiler">ResourceProfiler</a>
and profiling task streams with <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-local.html#profiler">Profiler</a> are other
commonly used local diagnostics.</p>
<p>For similar monitoring in a distributed cluster, see <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-distributed.html">Dask distributed diagnostics</a>.</p>
</div>
<div class="section" id="visualizing-computations">
<h3><a class="toc-backref" href="#id22">Visualizing computations</a><a class="headerlink" href="#visualizing-computations" title="Permalink to this headline">¶</a></h3>
<p>Dask allows you to <a class="reference external" href="https://docs.dask.org/en/latest/graphviz.html">visualize the task graph</a> of a computation
before running it, which can be handy when trying to understand where the bottlenecks are.</p>
<p>In most cases the number of tasks is too large to visualize, so it’s useful to restrict
the graph just a few chunks, as shown in this example.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [23]: </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>

<span class="gp">In [24]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=.</span><span class="mi">1</span><span class="p">)</span>

<span class="go"># Rechunk to illustrate multiple tasks</span>
<span class="gp">In [25]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">chunk</span><span class="p">({</span><span class="s2">&quot;variants&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">,</span> <span class="s2">&quot;samples&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">})</span>

<span class="gp">In [26]: </span><span class="n">counts</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_call_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span><span class="o">.</span><span class="n">call_allele_count</span><span class="o">.</span><span class="n">data</span>

<span class="go"># Restrict to first 3 chunks in variants dimension</span>
<span class="gp">In [27]: </span><span class="n">counts</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[:</span><span class="mi">3</span><span class="o">*</span><span class="n">counts</span><span class="o">.</span><span class="n">chunksize</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="o">...</span><span class="p">]</span>

<span class="gp">In [28]: </span><span class="n">counts</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">optimize_graph</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gh">Out[28]: </span><span class="go">&lt;IPython.core.display.Image object&gt;</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/mydask.png"><img alt="_images/mydask.png" class="align-center" src="_images/mydask.png" style="width: 600px;" /></a>
<p>By passing keyword arguments to <code class="docutils literal notranslate"><span class="pre">visualize</span></code> we can see the order tasks will run in:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># Graph where colors indicate task ordering</span>
<span class="gp">In [29]: </span><span class="n">counts</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;order&quot;</span><span class="p">,</span> <span class="n">optimize_graph</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;order&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;autumn&quot;</span><span class="p">,</span> <span class="n">node_attr</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;penwidth&quot;</span><span class="p">:</span> <span class="s2">&quot;4&quot;</span><span class="p">})</span>
<span class="gh">Out[29]: </span><span class="go">&lt;IPython.core.display.Image object&gt;</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/order.png"><img alt="_images/order.png" class="align-center" src="_images/order.png" style="width: 600px;" /></a>
<p>Task order number is shown in circular boxes, colored from red to yellow.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="vcf.html" class="btn btn-neutral float-right" title="Reading VCF" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="getting_started.html" class="btn btn-neutral float-left" title="Getting Started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, sgkit developers.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>