

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>User Guide &mdash; sgkit  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Reading VCF" href="vcf.html" />
    <link rel="prev" title="sgkit.variables.window_stop_spec" href="generated/sgkit.variables.window_stop_spec.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> sgkit
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API reference</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#io">IO</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#plink">PLINK</a></li>
<li class="toctree-l3"><a class="reference internal" href="#vcf">VCF</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#converting-genetic-data-to-zarr">Converting genetic data to Zarr</a></li>
<li class="toctree-l2"><a class="reference internal" href="#working-with-cloud-native-data">Working with cloud-native data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#deploying-sgkit-on-a-cluster">Deploying sgkit on a cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="#adding-custom-data-to-a-dataset">Adding custom data to a Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pca">PCA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#using-gpus">Using GPUs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#custom-computations">Custom Computations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="vcf.html">Reading VCF</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to sgkit</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sgkit</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>User Guide</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/user_guide.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="user-guide">
<h1>User Guide<a class="headerlink" href="#user-guide" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="table-of-contents">
<p class="topic-title">Table of contents:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#io" id="id2">IO</a></p>
<ul>
<li><p><a class="reference internal" href="#plink" id="id3">PLINK</a></p></li>
<li><p><a class="reference internal" href="#vcf" id="id4">VCF</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#converting-genetic-data-to-zarr" id="id5">Converting genetic data to Zarr</a></p></li>
<li><p><a class="reference internal" href="#working-with-cloud-native-data" id="id6">Working with cloud-native data</a></p></li>
<li><p><a class="reference internal" href="#deploying-sgkit-on-a-cluster" id="id7">Deploying sgkit on a cluster</a></p></li>
<li><p><a class="reference internal" href="#adding-custom-data-to-a-dataset" id="id8">Adding custom data to a Dataset</a></p></li>
<li><p><a class="reference internal" href="#pca" id="id9">PCA</a></p></li>
<li><p><a class="reference internal" href="#using-gpus" id="id10">Using GPUs</a></p></li>
<li><p><a class="reference internal" href="#custom-computations" id="id11">Custom Computations</a></p></li>
</ul>
</div>
<div class="section" id="io">
<h2><a class="toc-backref" href="#id2">IO</a><a class="headerlink" href="#io" title="Permalink to this headline">¶</a></h2>
<div class="section" id="plink">
<h3><a class="toc-backref" href="#id3">PLINK</a><a class="headerlink" href="#plink" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="generated/sgkit.io.plink.read_plink.html#sgkit.io.plink.read_plink" title="sgkit.io.plink.read_plink"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.plink.read_plink()</span></code></a> function loads a single PLINK dataset as Dask
arrays within an <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.16.2.dev3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> from <code class="docutils literal notranslate"><span class="pre">bed</span></code>, <code class="docutils literal notranslate"><span class="pre">bim</span></code>, and <code class="docutils literal notranslate"><span class="pre">fam</span></code> files.</p>
<p>PLINK IO support is an “extra” feature within sgkit and requires additional
dependencies. To install sgkit with PLINK support using pip:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install &#39;sgkit[plink]&#39;
</pre></div>
</div>
</div>
<div class="section" id="vcf">
<h3><a class="toc-backref" href="#id4">VCF</a><a class="headerlink" href="#vcf" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a> function converts one or more VCF files to
Zarr files stored in sgkit’s Xarray data representation, which can then be opened
as a <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.16.2.dev3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a>.</p>
<p>See <a class="reference internal" href="vcf.html#vcf"><span class="std std-ref">Reading VCF</span></a> for installation instructions, and details on using VCF in sgkit.</p>
</div>
</div>
<div class="section" id="converting-genetic-data-to-zarr">
<h2><a class="toc-backref" href="#id5">Converting genetic data to Zarr</a><a class="headerlink" href="#converting-genetic-data-to-zarr" title="Permalink to this headline">¶</a></h2>
<p>TODO: Describe the process and motivation for converting genetic file formats to Zarr prior to analysis</p>
</div>
<div class="section" id="working-with-cloud-native-data">
<h2><a class="toc-backref" href="#id6">Working with cloud-native data</a><a class="headerlink" href="#working-with-cloud-native-data" title="Permalink to this headline">¶</a></h2>
<p>TODO: Show how to read/write Zarr (and VCF?) data in cloud storage</p>
</div>
<div class="section" id="deploying-sgkit-on-a-cluster">
<h2><a class="toc-backref" href="#id7">Deploying sgkit on a cluster</a><a class="headerlink" href="#deploying-sgkit-on-a-cluster" title="Permalink to this headline">¶</a></h2>
<p>TODO: Create a tutorial on running sgkit at scale</p>
</div>
<div class="section" id="adding-custom-data-to-a-dataset">
<h2><a class="toc-backref" href="#id8">Adding custom data to a Dataset</a><a class="headerlink" href="#adding-custom-data-to-a-dataset" title="Permalink to this headline">¶</a></h2>
<p>TODO:  Show how something like sample metadata can be joined to an existing Xarray dataset. Also briefly explain
indexing and uniqueness within Xarray/Pandas, since this is critical for understanding joins.</p>
</div>
<div class="section" id="pca">
<h2><a class="toc-backref" href="#id9">PCA</a><a class="headerlink" href="#pca" title="Permalink to this headline">¶</a></h2>
<p>TODO: Describe the upstream tools for PCA (i.e. those in dask-ml/scikit-learn)</p>
</div>
<div class="section" id="using-gpus">
<h2><a class="toc-backref" href="#id10">Using GPUs</a><a class="headerlink" href="#using-gpus" title="Permalink to this headline">¶</a></h2>
<p>TODO: Show CuPy examples</p>
</div>
<div class="section" id="custom-computations">
<span id="id1"></span><h2><a class="toc-backref" href="#id11">Custom Computations</a><a class="headerlink" href="#custom-computations" title="Permalink to this headline">¶</a></h2>
<p>TODO: Finish explaining how Numba works and how users might apply it</p>
<p>Here is an example that demonstrates an alt allele count:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span> <span class="nn">numba</span>

<span class="gp">In [2]: </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>

<span class="gp">In [3]: </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="gp">In [4]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=.</span><span class="mi">2</span><span class="p">)</span>

<span class="gp">In [5]: </span><span class="k">def</span> <span class="nf">alt_allele_count</span><span class="p">(</span><span class="n">gt</span><span class="p">):</span>
<span class="gp">   ...: </span>    <span class="n">out</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">shape</span><span class="p">[:</span><span class="mi">2</span><span class="p">],</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
<span class="gp">   ...: </span>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">ndindex</span><span class="p">(</span><span class="o">*</span><span class="n">out</span><span class="o">.</span><span class="n">shape</span><span class="p">):</span>
<span class="gp">   ...: </span>        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">gt</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">):</span>
<span class="gp">   ...: </span>            <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">gt</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
<span class="gp">   ...: </span>    <span class="k">return</span> <span class="n">out</span>
<span class="gp">   ...: </span>

<span class="gp">In [6]: </span><span class="n">numba</span><span class="o">.</span><span class="n">njit</span><span class="p">(</span><span class="n">alt_allele_count</span><span class="p">)(</span><span class="n">ds</span><span class="o">.</span><span class="n">call_genotype</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="gh">Out[6]: </span><span class="go"></span>
<span class="go">array([[ 0,  1,  1],</span>
<span class="go">       [ 1,  1, -1],</span>
<span class="go">       [-1,  1,  2],</span>
<span class="go">       [ 0, -1, -1],</span>
<span class="go">       [ 1,  2,  2]])</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="vcf.html" class="btn btn-neutral float-right" title="Reading VCF" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="generated/sgkit.variables.window_stop_spec.html" class="btn btn-neutral float-left" title="sgkit.variables.window_stop_spec" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, sgkit developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>