

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>sgkit.gwas_linear_regression &mdash; sgkit  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="sgkit.create_genotype_call_dataset" href="sgkit.create_genotype_call_dataset.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> sgkit
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">API reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../api.html#top-level-functions">Top-level functions</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="sgkit.create_genotype_call_dataset.html">sgkit.create_genotype_call_dataset</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">sgkit.gwas_linear_regression</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">sgkit</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../api.html">API reference</a> &raquo;</li>
        
      <li>sgkit.gwas_linear_regression</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/generated/sgkit.gwas_linear_regression.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="sgkit-gwas-linear-regression">
<h1>sgkit.gwas_linear_regression<a class="headerlink" href="#sgkit-gwas-linear-regression" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="sgkit.gwas_linear_regression">
<code class="sig-prename descclassname">sgkit.</code><code class="sig-name descname">gwas_linear_regression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ds</span><span class="p">:</span> <span class="n">xarray.core.dataset.Dataset</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">dosage</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">covariates</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>Sequence<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">traits</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>Sequence<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">add_intercept</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span> &#x2192; xarray.core.dataset.Dataset<a class="headerlink" href="#sgkit.gwas_linear_regression" title="Permalink to this definition">¶</a></dt>
<dd><p>Run linear regression to identify continuous trait associations with genetic variants.</p>
<p>This method solves OLS regressions for each variant simultaneously and reports
effect statistics as defined in [1]. This is facilitated by the removal of
sample (i.e. person/individual) covariates through orthogonal projection
of both the genetic variant and phenotype data [2]. A consequence of this
rotation is that effect sizes and significances cannot be reported for
covariates, only variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ds</strong> (<em>Dataset</em>) – Dataset containing necessary dependent and independent variables.</p></li>
<li><p><strong>dosage</strong> (<em>str</em>) – Dosage variable name where “dosage” array can contain represent
one of several possible quantities, e.g.:
- Alternate allele counts
- Recessive or dominant allele encodings
- True dosages as computed from imputed or probabilistic variant calls
- Any other custom encoding in a user-defined variable</p></li>
<li><p><strong>covariates</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>Sequence</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Covariate variable names, must correspond to 1 or 2D dataset
variables of shape (samples[, covariates]). All covariate arrays
will be concatenated along the second axis (columns).</p></li>
<li><p><strong>traits</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>Sequence</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Trait (e.g. phenotype) variable names, must all be continuous and
correspond to 1 or 2D dataset variables of shape (samples[, traits]).
2D trait arrays will be assumed to contain separate traits within columns
and concatenated to any 1D traits along the second axis (columns).</p></li>
<li><p><strong>add_intercept</strong> (<em>bool</em><em>, </em><em>optional</em>) – Add intercept term to covariate set, by default True.</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Regression statistics from this implementation are only valid when an
intercept is present. The <cite>add_intercept</cite> flag is a convenience for adding one
when not already present, but there is currently no parameterization for
intercept-free regression.</p>
<p>Additionally, both covariate and trait arrays will be rechunked to have blocks
along the sample (row) dimension but not the column dimension (i.e.
they must be tall and skinny).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><dl class="simple">
<dt>Dataset containing (N = num variants, O = num traits):</dt><dd><dl class="simple">
<dt>beta<span class="classifier">(N, O) array-like</span></dt><dd><p>Beta values associated with each variant and trait</p>
</dd>
<dt>t_value<span class="classifier">(N, O) array-like</span></dt><dd><p>T statistics for each beta</p>
</dd>
<dt>p_value<span class="classifier">(N, O) array-like</span></dt><dd><p>P values as float in [0, 1]</p>
</dd>
</dl>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="http://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.15.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Regression statistics from this implementation are only valid when an
intercept is present. The <cite>add_intercept</cite> flag is a convenience for adding one
when not already present, but there is currently no parameterization for
intercept-free regression.</p>
<p>Additionally, both covariate and trait arrays will be rechunked to have blocks
along the sample (row) dimension but not the column dimension (i.e.
they must be tall and skinny).</p>
</div>
<p class="rubric">References</p>
<ul class="simple">
<li><dl class="simple">
<dt>[1] Hastie, Trevor, Robert Tibshirani, and Jerome Friedman. 2009. The Elements</dt><dd><p>of Statistical Learning: Data Mining, Inference, and Prediction, Second Edition.
Springer Science &amp; Business Media.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>[2] Loh, Po-Ru, George Tucker, Brendan K. Bulik-Sullivan, Bjarni J. Vilhjálmsson,</dt><dd><p>Hilary K. Finucane, Rany M. Salem, Daniel I. Chasman, et al. 2015. “Efficient
Bayesian Mixed-Model Analysis Increases Association Power in Large Cohorts.”
Nature Genetics 47 (3): 284–90.</p>
</dd>
</dl>
</li>
</ul>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="sgkit.create_genotype_call_dataset.html" class="btn btn-neutral float-left" title="sgkit.create_genotype_call_dataset" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, sgkit developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>